### PR TITLE
[GraphBolt] enable slice for integer-init ItemSet

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/itemset.py [hunk 1]
```diff
@@ -151,11 +151,10 @@ def __getitem__(self, idx: Union[int, slice, Iterable]) -> Tuple:
                 f"{type(self).__name__} instance doesn't support indexing."
             )
         if isinstance(self._items, int):
-            assert isinstance(idx, (int, torch.Tensor)), (
-                f"Indexing of integer-initialized {type(self).__name__} "
-                f"instance must be int or torch.Tensor."
-            )
             # [Warning] Index range is not checked.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/itemset.py:154]: Does this warning still apply?

Comment (Rhett-Ying) [reply python/dgl/graphbolt/itemset.py:154]: yes, it's applied to `return idx`

Comment (frozenbugs) [reply python/dgl/graphbolt/itemset.py:154]: How hard to apply an filter?

Comment (Rhett-Ying) [reply python/dgl/graphbolt/itemset.py]: added for `int` and remove support for `Tensor` as `slice` and `int` are the ways we index in `ItemSampler` for now. If we need more indexing ways, we enhance later. Adding support `Tensor` requires more checks such as range, dtype.

File: python/dgl/graphbolt/itemset.py [hunk 2]
```diff
@@ -151,12 +151,20 @@ def __getitem__(self, idx: Union[int, slice, Iterable]) -> Tuple:
...
-            assert isinstance(idx, (int, torch.Tensor)), (
-                f"Indexing of integer-initialized {type(self).__name__} "
-                f"instance must be int or torch.Tensor."
+            if isinstance(idx, slice):
+                start, stop, step = idx.indices(self._items)
+                return torch.arange(start, stop, step)
+            if isinstance(idx, int):
+                if idx < 0:
+                    idx += self._items
+                if idx >= self._items:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/itemset.py:160]: Should we check idx < 0: again here?

File: tests/python/pytorch/graphbolt/test_itemset.py [hunk 3]
```diff
@@ -115,8 +115,17 @@ def test_ItemSet_seed_nodes():
         assert i == item_set[i]
     # Indexing with a slice.
     assert torch.equal(item_set[:], torch.arange(0, 5))
-    # Indexing with an Tensor.
-    assert torch.equal(item_set[torch.arange(0, 5)], torch.arange(0, 5))
+    # Indexing with an integer.
+    assert item_set[0] == 0
+    assert item_set[-1] == 4
+    # Indexing that is out of range.
+    with pytest.raises(IndexError, match="ItemSet index out of range."):
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_itemset.py:122]: Check item_set[-100] as well?

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/itemset.py (hunk 2 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/itemset.py:160]: yes

File: tests/python/pytorch/graphbolt/test_itemset.py (hunk 3 above)
Comment (Rhett-Ying) [reply tests/python/pytorch/graphbolt/test_itemset.py:122]: added.

**Pull request merged**